### PR TITLE
feat: session cleanup after submission

### PR DESCRIPTION
**Pull request by franciscog-kainos**
### JIRA link
https://companieshouse.atlassian.net/browse/LFA-1641

### Change description
After checking the penalty details the user is not able to go back to change them. The only permitted route is to go back to the beginning to the journey by pressing the govuk icon.
**Note:** The user can still access the confirmation screen 
image

### Work checklist

- [x] Tests added where applicable
- [x] UI changes look good on mobile
- [x] UI changes meet accessibility criteria

### Merge instructions

We are commited to keeping commit history clean, consistent and linear. To achive this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/models/ApplicationData.ts [hunk 1]
```diff
@@ -7,4 +7,5 @@ export const APPLICATION_DATA_KEY = 'appeals';
 export interface ApplicationData {
     appeal: Appeal;
     navigation: Navigation;
+    submittedAppeal?: Appeal;
```
Comment (akaJSaunders) [inline src/models/ApplicationData.ts:10]: Why store another appeal? Couldn't we just use a flag that returns true as suggested by the spike?

Comment (franciscog-kainos) [reply src/models/ApplicationData.ts:10]: The flag in itself would mean that we would have to somehow retrieve the appeal from the api. I took the best of both solutions described in the spike and implemented this instead. No required api calls and simple implementation

Comment (franciscog-kainos) [reply src/models/ApplicationData.ts:10]: Also, we are not storing another appeal, we are simply moving the submitted appeal to a 'confirmed' stage while we clear the 'build' stage for the new appeal

File: src/controllers/CheckYourAppealController.ts [hunk 2]
```diff
@@ -39,7 +40,7 @@ const navigation = {
 export class CheckYourAppealController extends SafeNavigationBaseController<any> {
     constructor() {
         super(template, navigation, undefined, undefined, [AppealStorageFormActionProcessor,
-            InternalEmailFormActionProcessor, UserEmailFormActionProcessor]);
+            InternalEmailFormActionProcessor, UserEmailFormActionProcessor, SessionCleanupProcessor]);
```
Comment (akaJSaunders) [inline src/controllers/CheckYourAppealController.ts:43]: why are we setting
`appData.submittedAppeal = appData.appeal`
at the check-your-appeal page? An appeal has not been submitted at this point

Comment (franciscog-kainos) [reply src/controllers/CheckYourAppealController.ts:43]: At the point of execution of the SessionCleanupProcessor, which is after the appeal has been submitted by the AppealStorageFormActionProcessor, we can safely say that the appeal has been submitted to mongo and therefore we can mark this appeal as 'complete' by moving it to the `submittedAppeal` field. And deleting the current `appeal` from the session

File: src/controllers/processors/SessionCleanupProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import { Session } from 'ch-node-session-handler';
+import { Request } from 'express';
+import { provide } from 'inversify-binding-decorators';
+
+import { FormActionProcessor } from 'app/controllers/processors/FormActionProcessor';
+import { ApplicationData, APPLICATION_DATA_KEY } from 'app/models/ApplicationData';
+import { APPLICATION_DATA_UNDEFINED, SESSION_NOT_FOUND_ERROR } from 'app/utils/CommonErrors';
+import { CONFIRMATION_PAGE_URI } from 'app/utils/Paths';
+
+@provide(SessionCleanupProcessor)
```
Comment (akaJSaunders) [inline src/controllers/processors/SessionCleanupProcessor.ts:10]: Maybe add one info/debug logging to inform that session has been cleaned up. What do you think?

File: src/controllers/CheckYourAppealController.ts (hunk 2 above)
Comment (akaJSaunders) [reply src/controllers/CheckYourAppealController.ts:43]: ah my bad, my head was in middleware mode, the processor makes more sense

File: test/controllers/CheckYourAppealController.test.ts [hunk 4]
```diff
@@ -210,17 +241,22 @@ describe('CheckYourAppealController', () => {
                 'xUHinh19D17SQV2BYRLnGEZgeovYhcVitzLJMxpGxXW0w+30EYBb+6yF44pDWPsPejI17R5JSwy/Cw5kYQKO2A==';
 
             const appealsService = createSubstituteOf<AppealsService>(service => {
-                service.save(Arg.any(), Arg.any(), Arg.any()).returns(Promise.resolve(Arg.any()));
+                service.save(Arg.any(), Arg.any(), Arg.any()).returns(Promise.resolve('1'));
```
Comment (akaJSaunders) [inline test/controllers/CheckYourAppealController.test.ts:244]: Why did you change this to 1?

File: src/controllers/processors/SessionCleanupProcessor.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import { FormActionProcessor } from 'app/controllers/processors/FormActionProcessor';
+import { loggerInstance } from 'app/middleware/Logger';
+import { ApplicationData, APPLICATION_DATA_KEY } from 'app/models/ApplicationData';
+import { APPLICATION_DATA_UNDEFINED, SESSION_NOT_FOUND_ERROR } from 'app/utils/CommonErrors';
+import { CONFIRMATION_PAGE_URI } from 'app/utils/Paths';
+
+@provide(SessionCleanupProcessor)
+export class SessionCleanupProcessor implements FormActionProcessor {
+
+    public async process(req: Request): Promise<void> {
```
Comment (akaJSaunders) [inline src/controllers/processors/SessionCleanupProcessor.ts:14]: I think we need some unit tests for this processor

Comment (franciscog-kainos) [reply src/controllers/processors/SessionCleanupProcessor.ts:14]: I don't think we need that since all the logic was tested implicitly by the CheckYourAppealController.test.ts.

File: test/controllers/CheckYourAppealController.test.ts (hunk 4 above)
Comment (franciscog-kainos) [reply test/controllers/CheckYourAppealController.test.ts:244]: Substitute was having a hard time comparing Arg.any() when we do 
image

Comment (franciscog-kainos) [reply test/controllers/CheckYourAppealController.test.ts:244]: The save method was saying that the args didn't match because { id: Arg.any() } did not equal { id: Arg.any() } :/

File: src/controllers/processors/SessionCleanupProcessor.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+        if (!appData) {
+            throw APPLICATION_DATA_UNDEFINED;
+        }
+
+        appData.submittedAppeal = appData.appeal;
+
+        delete appData.appeal;
+
+        appData.navigation.permissions = [CONFIRMATION_PAGE_URI];
```
Comment (akaJSaunders) [inline src/controllers/processors/SessionCleanupProcessor.ts:32]: I understand why you are doing this, but what if a user submits an appeal, and then starts a new one, doesn't this mean they will have permissions to jump to the confirmation page at any time even though they might not have completed the new journey

Comment (franciscog-kainos) [reply src/controllers/processors/SessionCleanupProcessor.ts:32]: Yes they will be able to see the appeal they just submitted. Sounds like a feature to me.

Comment (franciscog-kainos) [reply src/controllers/processors/SessionCleanupProcessor.ts:32]: Harmless feature

File: src/controllers/CheckYourAppealController.ts (hunk 2 above)
Comment (n-r-b) [reply src/controllers/CheckYourAppealController.ts:43]: Let's say the customer has submitted an appeal, and they navigate back to the start of the journey, will their stale appeal data from their active session populate the fields on the `appeal-a-penalty/penalty-reference` page? Or does this get cleared before they get to the summary page?

File: src/controllers/processors/SessionCleanupProcessor.ts (hunk 6 above)
Comment (akaJSaunders) [reply src/controllers/processors/SessionCleanupProcessor.ts:32]: True, I guess you are using the appealSubmitted data for that
